bpf: initial pcap exporter for lb

borkmann opened the pull request:
See commit msgs.

Basis for agent + Hubble recorder framework:
 - https://github.com/cilium/cilium/pull/15497
 - https://github.com/cilium/cilium/pull/15385

borkmann committed:
bpf: add per-cpu ktime cache for faster subsequent access

Add a per-cpu ktime cache we can use for the packet capturing exporter
facility from the LB.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add initial infra for exporting pcap header into perf RB

Add a new capture type for pcap exporter to perf RB that we're going to use
for the lb-only mode out of XDP initially.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: hook up cilium_capture_{in,out} helpers to lb-only + dsr

Add both to the XDP DSR load balancer for capture points upon
packet arrival and departure after encap.

Also guard it under ENABLE_CAPTURE so that there's no overhead
when not compiled in, plus so that the agent can probe the boot
ktime helper availability.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add rule_id parameter to cilium_capture_{in,out} helpers

And also add __ prefixed variants that abstract the ktime handling
away from caller. Right now the capture is unconditional, next step
we'll add a classifier where the value (rule_id) of the match is
then passed through the capture helpers.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: move capturing behind an elf templating knob

If there are no capturing rules from agent side, then agent emits the
capture_enabled ELF var as 0 so that verifier's dead code elimination
will wire up the branches statically. Once there is one or more users
(rules) then capture_enabled ELF var is rewritten as 1 and reloaded,
and similarly when all rules have been removed, the agent rewrites the
capture_enabled ELF var to 0 again.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add classification cache to avoid redissection

Idea is to only dissect and lookup a packet once, and cache the result
such that after DSR encap we do not need to look into the packet again.
Instead we can just rely on a simple per-CPU cache.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add classification key & mask struct

Add a v4 hash table map as well as key (mask) and value structures. The
idea is to emulate a poor man's wildcard flow lookup via hash table map
until we have a proper wildcard map support in the kernel. The rule_id
will be passed into the perf RB so that the user space agent can pick
it up and correlate the matched rule.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: implement capture wildcard v4 classifier

Implement the wild-carded v4 classifier part. The agent is generating and
emitting the PREFIX_MASKS4 and regenerating if a mask was added or removed.
The cilium_capture4_rules can have n entries with m different PREFIX_MASKS4
where n >> m. Lookup performance depends mainly on m. In here is a fallback /
example definition mainly as example and for compile testing, given agent
typically emits this instead.

To scale more combinations in future we need a new BPF map implementation
in the kernel based on algorithms like TupleMerge [0].

  [0] http://www.cse.msu.edu/~dalyjame/papers/TupleMerge.pdf

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add option to specify cap len for pcap exporter

Allow to specify a max capture length for the BPF map, so that a user is
able to configure this instead of always full length.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: add agent switch to enable pcap recorder

Add an --enable-recorder switch for the agent to opt into the pcap exporter
which then checks underlying kernel dependencies and compiles in the code
into LB XDP fast path.

Example agent:

  ./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=true       \
      --datapath-mode=lb-only --bpf-lb-algorithm=maglev             \
      --bpf-lb-maglev-table-size=2039 --bpf-lb-mode=dsr             \
      --bpf-lb-acceleration=native --devices=enp10s0f0np0           \
      --bpf-lb-dsr-dispatch=ipip --disable-envoy-version-check=true \
      --enable-bpf-clock-probe=true --enable-recorder=true

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: add wildcarded capture map align checker

Add wildcarded capture map align checker bits, so that BPF vs Go representation
can be asserted.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: add wildcarded capture map iface for dumper

Add a common interface such that we prep both v4/v6 maps for allowing to
dump their entries from CLI.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: add wildcarded capture map representation

Add a basic v4/v6 map representation so that the user space bits can build
on top of it. This is needed for later map+mask management as well as a
Cilium CLI map dump for introspection. Default map size is currently fixed,
but we can add an agent knob at some point if needed.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: implement capture wildcard v6 classifier

Implement the wild-carded v6 classifier part. The implementation follow the
v4 bits and pieces.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: add bpf recorder list command to dump wildcarded entries

Add `cilium bpf recorder list` command to dump both the v4 and v6 wildcarded
maps tht contain the capture n-tuple filters.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, doc: add cmdref doc entries for cilium bpf recorder

Add generated cmdref to the documentation.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, bugtool: add bpf recorder dump to sysdump

Add the dump of the recorder map to the sysdump for introspection.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, monitor: add recorder capture monitoring event

Allow for easy debugging of recorder events and thus provide a monitoring
dissector.

Example:

  # ./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=true       \
        --datapath-mode=lb-only --bpf-lb-algorithm=maglev             \
        --bpf-lb-maglev-table-size=2039 --bpf-lb-mode=dsr             \
        --bpf-lb-acceleration=native --devices=enp10s0f0np0           \
        --bpf-lb-dsr-dispatch=ipip --disable-envoy-version-check=true \
        --enable-bpf-clock-probe=true --enable-recorder=true

  # bpftool map update pinned /sys/fs/bpf/tc/globals/cilium_capture4_rules \
        key hex 0 0 0 0 c0 a8 a0 04 0 0 0 0 0 0 20 0 value hex 0 1 0 0 0 0 0 0

  # ./cilium/cilium bpf recorder list
  192.168.160.4/32:0 -> 0.0.0.0/0:0 ANY ID:256 CapLen:0

  # ./cilium/cilium service list
  ID   Frontend             Service Type   Backend
  1    192.168.160.3:8080   ExternalIPs    1 => 192.168.160.4:8080

  (from remote)# curl 192.168.160.3:8080
  [...]

  # ./cilium/cilium monitor
  [...]
  Recorder capture: dir:ingress rule:256 ts:99992076120464 caplen:74 len:74
  Ethernet	{Contents=[..14..] Payload=[..62..] SrcMAC=b8:ce:f6:05:e7:76 DstMAC=b8:ce:f6:05:e7:62 EthernetType=IPv4 Length=0}
  IPv4	{Contents=[..20..] Payload=[..40..] Version=4 IHL=5 TOS=0 Length=60 Id=44495 Flags=DF FragOffset=0 TTL=64 Protocol=TCP Checksum=52115 SrcIP=192.168.160.4 DstIP=192.168.160.3 Options=[] Padding=[]}
  TCP	{Contents=[..40..] Payload=[] SrcPort=52128 DstPort=8080(http-alt) Seq=592744369 Ack=0 DataOffset=10 FIN=false SYN=true RST=false PSH=false ACK=false URG=false ECE=false CWR=false NS=false Window=42340 Checksum=59261 Urgent=0 Options=[..5..] Padding=[]}
  ----
  Recorder capture: dir:egress rule:256 ts:99992076120464 caplen:94 len:94
  Ethernet	{Contents=[..14..] Payload=[..86..] SrcMAC=b8:ce:f6:05:e7:62 DstMAC=b8:ce:f6:05:e7:76 EthernetType=IPv4 Length=0}
  IPv4	{Contents=[..20..] Payload=[..40..] Version=4 IHL=5 TOS=0 Length=60 Id=44495 Flags=DF FragOffset=0 TTL=64 Protocol=TCP Checksum=52115 SrcIP=192.168.160.4 DstIP=192.168.160.3 Options=[] Padding=[]}
  IPv4	{Contents=[..20..] Payload=[..40..] Version=4 IHL=5 TOS=0 Length=60 Id=44495 Flags=DF FragOffset=0 TTL=64 Protocol=TCP Checksum=52115 SrcIP=192.168.160.4 DstIP=192.168.160.3 Options=[] Padding=[]}
  TCP	{Contents=[..40..] Payload=[] SrcPort=52128 DstPort=8080(http-alt) Seq=592744369 Ack=0 DataOffset=10 FIN=false SYN=true RST=false PSH=false ACK=false URG=false ECE=false CWR=false NS=false Window=42340 Checksum=59261 Urgent=0 Options=[..5..] Padding=[]}
  ----
  [...]

Note that our monitor dissector code is currently broken in that it cannot
handle IPIP correctly (hence the same IP header twice above). It can only
cache/dump IP layer once right now, subject to a different fix at some point.

The future pcap writer code, needs to:

  1) Translate the boot time ts into a time of day ts.
  2) Dump into a different pcap file based on rule id.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

gandro commented on pkg/maps/recorder/ipv6.go:
```diff
@@ -0,0 +1,116 @@
+		k.DestAddr,
+		int(k.DestMask),
+		k.DestPort,
+		dumpNextHdr(int(k.NextHdr))))
```
nit:
```suggestion
		u8proto.U8proto(k.NextHdr)
```

Should allow you to get rid of `dumpNextHdr`

gandro commented on pkg/maps/recorder/ipv4.go:
```diff
@@ -0,0 +1,116 @@
+		k.DestAddr,
+		int(k.DestMask),
+		k.DestPort,
+		dumpNextHdr(int(k.NextHdr))))
```
nit:
```suggestion
		u8proto.U8proto(k.NextHdr)
```

gandro commented on pkg/maps/recorder/recorder.go:
```diff
@@ -0,0 +1,91 @@
+	default:
+		return "???"
+	}
+}
```
nit: This can be removed if you use `"github.com/cilium/cilium/pkg/u8proto"` instead (see above)

gandro commented on pkg/monitor/api/types.go:
```diff
@@ -50,6 +50,10 @@ const (
 
+	// MessageTypeRecCapture is a BPF datapath notification carrying a RecorderCapture
+	// which corresponds to capture_msg defined in bpf/lib/pcap.h
+	MessageTypeRecCapture
```
Please also add a `MessageTypeNameRecCapture` below (line 66+) and add it to `MessageTypeNames`

This the name (e.g. `record`) will allow you to filter it out in the CLI via e.g. `cilium monitor -t record` in the `m.match` statement you added in `pkg/monitor/format`

gandro commented on pkg/monitor/format/format.go:
```diff
@@ -141,6 +141,18 @@ func (m *MonitorFormatter) policyVerdictEvents(prefix string, data []byte) {
+	rc := monitor.RecorderCapture{}
+
+	if err := binary.Read(bytes.NewReader(data), byteorder.Native, &rc); err != nil {
+		fmt.Printf("Error while parsing policy notification message: %s\n", err)
```
```suggestion
		fmt.Printf("Error while parsing capture record: %s\n", err)
```

gandro reviewed (commented):
I did a rough pass over the monitor bits, looks good to me overall, a few minor comments though.

brb reviewed (approved):
Amazing!

borkmann committed:
cilium: don't pull in capture rules map if disabled

Move the v4/v6 maps under ENABLE_CAPTURE so they are not attempted to be
included given we also don't define their names/sizes from agent size.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
